library_generation/utils/generation_config_comparator.py

```diff
@@ -0,0 +1,324 @@
+    affected libraries.
+    :param baseline_library_configs: a list of LibraryConfig object
+    :param latest_library_configs: a list of LibraryConfig object
+    """
+    baseline_libraries = __convert_to_hashed_library_dict(baseline_library_configs)
+    latest_libraries = __convert_to_hashed_library_dict(latest_library_configs)
+    changed_libraries = []
+    # 1st round comparison.
+    for library_name, hash_library in baseline_libraries.items():
+        # 1. find any library removed from baseline_libraries.
+        # a library is removed from baseline_libraries if the library_name
+        # is not in latest_libraries.
+        if library_name not in latest_libraries:
+            if ChangeType.LIBRARIES_REMOVAL not in diff:
+                diff[ChangeType.LIBRARIES_REMOVAL] = []
+            diff[ChangeType.LIBRARIES_REMOVAL].append(library_name)
+        # 2. find any library that exists in both configs but at least one
+        # parameter is changed, which means the hash value is different.
+        if (
+            library_name in latest_libraries
+            and hash_library.hash_value != latest_libraries[library_name].hash_value
+        ):
+            changed_libraries.append(library_name)
+    # 2nd round comparison.
+    for library_name in latest_libraries:
+        # find any library added to latest_libraries.
+        # a library is added to latest_libraries if the library_name
+        # is not in baseline_libraries.
+        if library_name not in baseline_libraries:
+            if ChangeType.LIBRARIES_ADDITION not in diff:
+                diff[ChangeType.LIBRARIES_ADDITION] = []
+            diff[ChangeType.LIBRARIES_ADDITION].append(library_name)
+    # 3rd round comparison.
+    __compare_changed_libraries(
+        diff=diff,
+        baseline_libraries=baseline_libraries,
+        latest_libraries=latest_libraries,
+        changed_libraries=changed_libraries,
+    )
+
+
+def __convert_to_hashed_library_dict(
+    libraries: List[LibraryConfig],
+) -> Dict[str, HashLibrary]:
+    """
+    Convert a list of LibraryConfig objects to a Dict.
+    For each library object, the key is the library_name of the object, the
+    value is a HashLibrary object.
+
+    :param libraries: a list of LibraryConfig object.
+    :return: a mapping from library_name to HashLibrary object.
+    """
+    return {
+        get_library_name(library): HashLibrary(hash(library), library)
+        for library in libraries
+    }
+
+
+def __compare_changed_libraries(
+    diff: Dict[ChangeType, List[str]],
+    baseline_libraries: Dict[str, HashLibrary],
+    latest_libraries: Dict[str, HashLibrary],
+    changed_libraries: List[str],
+) -> None:
+    """
+    Compare each library with the same library_name to find what parameters are
+    changed.
+
+    Some parameters should not change:
+
+    - api_shortname
+    - library_name
+    - distribution_name
+    - group_id
```

Comment:
Technically anything can be changed, even including `library_name`. If `library_name` changes, we can treat it as a `removal of existing library + new client library`. And I have seen  service teams try to change `group_id` in the past. Is there challenges that prevent us from doing so? Or it's just we think they should not be changed logically?

Reply:
I need to have a key to determine whether a library is changed or removed/added and I choose the key to be `library_name`.

The logic is generic enough such that if `baseline_config` has a library, `a` (library_name), and `latest_config` has a library, `b` (library_name), then the comparison will be `a` is removed from `baseline_config` and `b` is added to `lastest_config`.

Reply:
I agree that the existing logic should be able to handle this. That's why I don't think we want to mention `library_name` can not be changed. For other things like `api_shortname`, I don't think we want to mention them either, because our scripts should just generate accordingly, is that right?

Reply:
IIUC, if the `library_name` or `api_shorname` is changed, we should treat this case as `LIBRARY_REMOVAL` and `LIBRARY_ADDITION`, right?

> And I have seen service teams try to change group_id in the past. Is there challenges that prevent us from doing so? Or it's just we think they should not be changed logically?

I don't know we have these cases. I think we can handle this case as distribution name change or group id change.

Reply:
Done, added unit tests to verify these parameters.